test/performance/io/format_fastq_benchmark.cpp

```diff
@@ -0,0 +1,242 @@
+
+// !\author: Clemens Cords <clemens.cords@fu-berlin.de>
+
+#include <benchmark/benchmark.h>
+#include <cstring>
+#include <fstream>
+
+#include <seqan3/test/seqan2.hpp>
+#if SEQAN3_HAS_SEQAN2
+#include <seqan/seq_io.h>
+#endif
+
+#include <seqan3/alphabet/nucleotide/dna5.hpp>
+#include <seqan3/alphabet/quality/phred42.hpp>
+#include <seqan3/io/sequence_file/format_fastq.hpp>
+#include <seqan3/io/sequence_file/input.hpp>
+#include <seqan3/range/view/to_char.hpp>
+#include <seqan3/std/filesystem>
+#include <seqan3/test/performance/sequence_generator.hpp>
+#include <seqan3/test/tmp_filename.hpp>
+
+using namespace seqan3;
+
+constexpr unsigned seed = 1234u;
+
+// ============================================================================
+// generate fastq file
+// ============================================================================
+constexpr size_t default_sequence_length = 50; //length of nucleotide and quality sequence
+
+std::string get_file(size_t n_entries)
+{
+    std::string file{};
+    std::string const id{"@name"};
+
+    for (size_t i = 0; i < n_entries; ++i)
+    {
+        auto seq = test::generate_sequence<dna5>(default_sequence_length, 0, seed);
+        std::string seq_string = seq | view::to_char;
+
+        auto qual = test::generate_sequence<phred42>(default_sequence_length, 0, seed);
+        std::string qual_string = qual | view::to_char;
+
+        file += id + '\n' + seq_string + '\n' + '+' + '\n' + qual_string + '\n';
+    }
+
+    return file;
+}
+
+// ============================================================================
+// write 3-line entry to stream as often as possible
+// ============================================================================
+void fastq_write_to_stream(benchmark::State & state)
+{
+    std::ostringstream ostream;
+
+    sequence_file_format_fastq format;
+    sequence_file_output_options options{};
+
+    std::string const id{"@name"};
+    auto seq = test::generate_sequence<dna5>(default_sequence_length, 0, seed);
+    auto qual = test::generate_sequence<phred42>(default_sequence_length, 0, seed);
+
+    for (auto _ : state)
+    {
+        format.write(ostream, options, seq, id, qual);
+    }
+}
+
+// ============================================================================
+// read dummy fastq file from a stream
+// ============================================================================
+void fastq_read_from_stream(benchmark::State & state)
+{
+    sequence_file_format_fastq format;
+    sequence_file_input_options<dna5, false> const options{};
+
+    std::string id{};
+    std::vector<dna5> seq{};
+    std::vector<phred42> qual{};
+
+    size_t n_entries_in_file = state.range(0);
+    auto file = get_file(n_entries_in_file);
+
+    for (auto _ : state)
+    {
+        std::istringstream istream{file};
+        // refilling streams introduces the same constant to
+        // both seqan3 and 2 benchmark runtime so they are still comparable
```

Comment:
Not sure how I feel about this comment. It it helpful or not? It shows that you thought hard about this, but it seems to be more like a mental note for yourself than a note for everyone.

Reply:
I shortened it but I think it's important because it's not immediately obvious to the reader why the creation of the file should be inside the benchmark for loop instead of outside since it introduces a significant constant to runtime